### PR TITLE
Use biobuf optimaze startspeed

### PR DESCRIPTION
**Pull request by elliotchenzichang**
Use bufio.Reader to read entry from disk. Bufio.Reader will read a block(default size is 4KB) to its ring buffer  previously and copy the buffer to you when you read a data from it. It can reduce syscall nums. Here is the benchmark result:
`BenchmarkRecovery-10            1 2288482750 ns/op 1156858488 B/op 14041579 allocs/op`
`BenchmarkRecovery-10            4  265742781 ns/op 152531440 B/op  1089555 allocs/op`

### CODE REVIEW

File: recovery_reader.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	fd, err := os.OpenFile(path, os.O_RDWR, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+	return &fileRecovery{
+		reader: bufio.NewReader(fd),
+	}, nil
+}
+
+func (fr *fileRecovery) readEntry() (e *Entry, err error) {
```
Comment (xujiajun) [inline recovery_reader.go:24]: Recovery is easy to be associated with recovery related to golang's panic. Is it inappropriate for this file name?
 In addition, this readEntry method is very similar to the readAt code in the datafile (a lot of codes are the same https://github.com/nutsdb/nutsdb/blob/master/datafile.go#L60)

Comment (elliotchenzichang) [reply recovery_reader.go:24]: fileRecovery means recover file to entries in memory, and the code of readEntry method is very similar to the readAt code in the datafile, its only change the interact between nutsdb and file.

Comment (xujiajun) [reply recovery_reader.go:24]: I understand what you mean, but don't repeat the code, can it be refactored?

Comment (elliotchenzichang) [reply recovery_reader.go:25]: have a little refactor, please review it when you have time.

Comment (xujiajun) [reply recovery_reader.go:25]: ok

**Comment by xujiajun** [comment]
test code

```go
package main

import (
	"fmt"
	"log"
	"time"

	"github.com/shirou/gopsutil/mem"
	"github.com/xujiajun/nutsdb"
)

var (
	db     *nutsdb.DB
	bucket = "bucket_iterator_demo"
)

func initDB() {
	db, _ = nutsdb.Open(
		nutsdb.DefaultOptions,
		nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
		nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
		nutsdb.WithSegmentSize(10*nutsdb.MB),
		//nutsdb.WithStartFileLoadingMode(nutsdb.FileIO),
	)
}

func main() {
	v, _ := mem.VirtualMemory()
	fmt.Printf("Mem : %v MB , Free: %v MB , Used:%v MB , Usage:%f%%\n", v.Total/1024/1024, v.Available/1024/1024, v.Used/1024/1024, v.UsedPercent)
	now := time.Now()
	initDB()
	fmt.Println(time.Since(now))
	//tx, err := db.Begin(true)
	//if err != nil {
	//	panic(err)
	//}
	//
	//for i := 0; i < 1000000; i++ {
	//	key := []byte("key_" + fmt.Sprintf("%03d", i))
	//	val := []byte("valvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalvalval__" + fmt.Sprintf("%03d", i))
	//	if err = tx.Put(bucket, key, val, nutsdb.Persistent); err != nil {
	//		// tx rollback
	//		tx.Rollback()
	//		fmt.Printf("rollback ok, err %v:", err)
	//	}
	//}
	//err = tx.Commit()
	//if err != nil {
	//	panic(err)
	//}
	////forward iteration
	//forwardIteration()
	////reverse iterative
	//reverseIterative()

	//read()
	v, _ = mem.VirtualMemory()
	fmt.Printf("Mem : %v MB , Free: %v MB , Used:%v MB , Usage:%f%%\n", v.Total/1024/1024, v.Available/1024/1024, v.Used/1024/1024, v.UsedPercent)
}

func read() {
	if err := db.View(
		func(tx *nutsdb.Tx) error {
			for i := 0; i < 100; i++ {
				key := []byte("key_" + fmt.Sprintf("%03d", i))
				entry, err := tx.Get(bucket, key)
				if err != nil {
					return err
				}
				fmt.Println("key, entry", string(entry.Key), string(entry.Value))
			}
			return nil
		}); err != nil {
		log.Fatal(err)
	}
}

func forwardIteration() {
	fmt.Println("--------begin forwardIteration--------")
	tx, err := db.Begin(false)
	iterator := nutsdb.NewIterator(tx, bucket, nutsdb.IteratorOptions{Reverse: false})
	i := 0
	for i < 10 {
		ok, err := iterator.SetNext()
		fmt.Println("ok, err", ok, err)
		fmt.Println("Key: ", string(iterator.Entry().Key))
		fmt.Println("Value: ", string(iterator.Entry().Value))
		fmt.Println()
		i++
	}
	err = tx.Commit()
	if err != nil {
		panic(err)
	}
	fmt.Println("--------end forwardIteration--------")
}

func reverseIterative() {
	fmt.Println("--------start reverseIterative--------")
	tx, err := db.Begin(false)
	iterator := nutsdb.NewIterator(tx, bucket, nutsdb.IteratorOptions{Reverse: true})
	i := 0
	for i < 10 {
		ok, err := iterator.SetNext()
		fmt.Println("ok, err", ok, err)
		fmt.Println("Key: ", string(iterator.Entry().Key))
		fmt.Println("Value: ", string(iterator.Entry().Value))
		fmt.Println()
		i++
	}
	err = tx.Commit()
	if err != nil {
		panic(err)
	}
	fmt.Println("--------end reverseIterative--------")
}

```

**Comment by xujiajun** [comment]
### WithStartFileLoadingMode nutsdb.MMap

```
db, _ = nutsdb.Open(
    nutsdb.DefaultOptions,
    nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
    nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
    nutsdb.WithSegmentSize(10*nutsdb.MB),
)



Mem : 16384 MB , Free: 5541 MB , Used:10842 MB , Usage:66.179276%
3.539617066s
Mem : 16384 MB , Free: 5048 MB , Used:11335 MB , Usage:69.188476%


Mem : 16384 MB , Free: 5164 MB , Used:11219 MB , Usage:68.479156%
3.609468108s
Mem : 16384 MB , Free: 4836 MB , Used:11547 MB , Usage:70.480633%


Mem : 16384 MB , Free: 4913 MB , Used:11470 MB , Usage:70.008802%
3.891852345s
Mem : 16384 MB , Free: 5330 MB , Used:11053 MB , Usage:67.462826%
```


### WithStartFileLoadingMode nutsdb.FileIO
```
db, _ = nutsdb.Open(
    nutsdb.DefaultOptions,
    nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
    nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
    nutsdb.WithSegmentSize(10*nutsdb.MB),
    nutsdb.WithStartFileLoadingMode(nutsdb.FileIO),
)
 

Mem : 16384 MB , Free: 4836 MB , Used:11547 MB , Usage:70.478940%
4.49763299s
Mem : 16384 MB , Free: 5267 MB , Used:11116 MB , Usage:67.852092%



Mem : 16384 MB , Free: 5713 MB , Used:10670 MB , Usage:65.127826%
3.486461714s
Mem : 16384 MB , Free: 5267 MB , Used:11116 MB , Usage:67.852521%


Mem : 16384 MB , Free: 5713 MB , Used:10670 MB , Usage:65.130448%
3.327743815s
Mem : 16384 MB , Free: 5205 MB , Used:11178 MB , Usage:68.226719%
```


### Use biobuf optimaze

```
db, _ = nutsdb.Open(
    nutsdb.DefaultOptions,
    nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
    nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
    nutsdb.WithSegmentSize(10*nutsdb.MB),
    //nutsdb.WithStartFileLoadingMode(nutsdb.FileIO),
)

Mem : 16384 MB , Free: 5489 MB , Used:10894 MB , Usage:66.494942%
1.521298024s
Mem : 16384 MB , Free: 5012 MB , Used:11371 MB , Usage:69.406962%


Mem : 16384 MB , Free: 5842 MB , Used:10541 MB , Usage:64.339900%
1.537471034s
Mem : 16384 MB , Free: 5283 MB , Used:11100 MB , Usage:67.750406%



Mem : 16384 MB , Free: 5680 MB , Used:10703 MB , Usage:65.326142%
1.53916012s
Mem : 16384 MB , Free: 5145 MB , Used:11238 MB , Usage:68.592930%

```
### In Conclusion

> After optimization, the performance is  1x faster than before. This is related to the size of the value. In my example, the value is more than 100 bytes. In theory, if the value is small, the startup will be faster.

**Comment by xujiajun** [comment]
## big value: size > 4096

### WithStartFileLoadingMode nutsdb.MMap
    db, _ = nutsdb.Open(
        nutsdb.DefaultOptions,
        nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
        nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
        nutsdb.WithSegmentSize(10*nutsdb.MB),
        //nutsdb.WithStartFileLoadingMode(nutsdb.FileIO),
    )

```

Mem : 16384 MB , Free: 5770 MB , Used:10613 MB , Usage:64.779758%
413.418707ms
Mem : 16384 MB , Free: 5484 MB , Used:10899 MB , Usage:66.524982%


Mem : 16384 MB , Free: 5371 MB , Used:11012 MB , Usage:67.217159%
434.746895ms
Mem : 16384 MB , Free: 5076 MB , Used:11307 MB , Usage:69.013977%


Mem : 16384 MB , Free: 5016 MB , Used:11367 MB , Usage:69.381046%
519.060424ms
Mem : 16384 MB , Free: 4791 MB , Used:11592 MB , Usage:70.757031%
```

### WithStartFileLoadingMode nutsdb.FileIO
```
    db, _ = nutsdb.Open(
        nutsdb.DefaultOptions,
        nutsdb.WithEntryIdxMode(nutsdb.HintKeyAndRAMIdxMode),
        nutsdb.WithDir("/tmp/nutsdbexample/example_iterator"),
        nutsdb.WithSegmentSize(10*nutsdb.MB),
        nutsdb.WithStartFileLoadingMode(nutsdb.FileIO),
    )


Mem : 16384 MB , Free: 5074 MB , Used:11309 MB , Usage:69.025540%
547.670726ms
Mem : 16384 MB , Free: 5226 MB , Used:11157 MB , Usage:68.097782%



Mem : 16384 MB , Free: 5305 MB , Used:11078 MB , Usage:67.615390%
436.40822ms
Mem : 16384 MB , Free: 5044 MB , Used:11339 MB , Usage:69.211984%

Mem : 16384 MB , Free: 5370 MB , Used:11013 MB , Usage:67.221189%
419.29026ms
Mem : 16384 MB , Free: 5092 MB , Used:11291 MB , Usage:68.918276%


```


## Use biobuf optimaze
```

Mem : 16384 MB , Free: 5302 MB , Used:11081 MB , Usage:67.639112%
421.394428ms
Mem : 16384 MB , Free: 5302 MB , Used:11081 MB , Usage:67.639112%



Mem : 16384 MB , Free: 5248 MB , Used:11135 MB , Usage:67.967439%
522.028629ms
Mem : 16384 MB , Free: 5086 MB , Used:11297 MB , Usage:68.955803%


Mem : 16384 MB , Free: 5295 MB , Used:11088 MB , Usage:67.680788%
467.958015ms
Mem : 16384 MB , Free: 5066 MB , Used:11317 MB , Usage:69.077158%
```


### In Conclusion

In the case of large value, the startup speed is about the same
